particle-iot/device-os
# Compressed/combined OTA updates

### Problem

This PR implements support for compressed/combined OTA updates on Gen 3 platforms.

Notes:

- A bootloader update is required.
- For now, only system and user part modules can be compressed/combined.
- This feature requires server-side support. We haven't yet deployed it.
- Unit tests are currently failing on CI due to missing zlib libraries. An update to the buildpack is pending.

### Steps to Test

1. Change the bootloader version in system_module_version.mk to 1000 (2.0.0), compile and flash the bootloader to a Gen 3 device.
2. Change the system firmware version to 1502 (1.5.0), compile and flash the system part and example app to the device.
3. Change the system firmware version to 1511 (1.5.1), recompile the app and flash it to the device OTA.
4. Verify via the log that the device received a compressed binary. Device should update from 1.5.0 to 1.5.1 without entering safe mode.

### Example App

```cpp
#include "application.h"

SYSTEM_MODE(SEMI_AUTOMATIC)
// SYSTEM_THREAD(ENABLED)

namespace {

system_tick_t updateStart = 0;

void firmwareUpdateEvent(system_event_t event, int param, void* data) {
    switch (param) {
    case firmware_update_begin: {
        const auto file = (FileTransfer::Descriptor*)data;
        Log.info("Firmware update started; file size: %uK", (unsigned)file->file_length / 1024);
        updateStart = millis();
        break;
    }
    case firmware_update_complete: {
        Log.info("Firmware update completed in %us", (unsigned)(millis() - updateStart) / 1000);
        break;
    }
    case firmware_update_failed: {
        Log.error("Firmware update failed");
        break;
    }
    default:
        break;
    }
}

void resetPendingEvent(system_event_t event, int param, void* data) {
    HAL_Delay_Milliseconds(1000);
}

const Serial1LogHandler logHandler(115200, LOG_LEVEL_WARN, {
    { "comm.ota", LOG_LEVEL_ALL },
    { "hal.ota", LOG_LEVEL_ALL },
    { "app", LOG_LEVEL_ALL }
});

} // namespace

void setup() {
    System.on(firmware_update, firmwareUpdateEvent);
    System.on(reset_pending, resetPendingEvent);
    Particle.connect();
}

void loop() {
}
```

### References

- [ch53174]
